subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,702 @@
+		 */
+		if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+			union pd_header header = prl_rx->emsg.header;
+
+			if (received_data_message(dev, header, PD_DATA_REQUEST)) {
+				/* Set to highest revision supported by both ports */
+				prl_set_rev(dev, PD_PACKET_SOP,
+					MIN(PD_REV30, header.specification_revision));
+				pe_set_state(dev, PE_SRC_NEGOTIATE_CAPABILITY);
+			}
+		}
+		/*
+		 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset
+		 * state when:
+		 *	1) The SenderResponseTimer times out
+		 */
+		else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+			pe_set_state(dev, PE_SRC_HARD_RESET);
+		}
+		break;
+	}
+}
+
+/**
+ * @brief 8.3.3.2.4 PE_SRC_Negotiate_Capability State
+ */
+void pe_src_negotiate_capability_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	LOG_INF("PE_SRC_Negotiate_Capability");
+
+	/* Get sink request */
+	pe->snk_request = *(uint32_t *)prl_rx->emsg.data;
+
+	/*
+	 * Ask the Device Policy Manager to evaluate the Request
+	 * from the Attached Sink.
+	 */
+	pe->snk_request_reply =
+		policy_check_sink_request(dev, pe->snk_request);
+
+	/*
+	 * The Policy Engine Shall transition to the
+	 * PE_SRC_Transition_Supply state when:
+	 *	1) The Request can be met.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
+		pe_set_state(dev, PE_SRC_TRANSITION_SUPPLY);
+	}
+	/*
+	 * The Policy Engine Shall transition to the
+	 * PE_SRC_Capability_Response state when:
+	 *	1) The Request cannot be met.
+	 *	2) Or the Request can be met later from the Power Reserve.
+	 */
+	else {
+		pe_set_state(dev, PE_SRC_CAPABILITY_RESPONSE);
+	}
+}
+
+/**
+ * @brief 8.3.3.2.5 PE_SRC_Transition_Supply State
+ */
+void pe_src_transition_supply_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	LOG_INF("PE_SRC_Transition_Supply");
+
+	/*
+	 * If snk_request_reply is set, this state was entered
+	 * from PE_SRC_Negotiate_Capability. So send Accept Message
+	 * and inform the Device Policy Manager that it Shall transition
+	 * the power supply to the Requested power level.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_ACCEPT);
+	}
+	/*
+	 * If snk_request_reply is not valid, this state was entered
+	 * from PE_SRC_Ready. So send GotoMin Message.
+	 */
+	else {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_GOTO_MIN);
+	}
+}
+
+void pe_src_transition_supply_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Ready state when:
+	 *	1) The Device Policy Manager informs the Policy Engine that
+	 *	   the power supply is ready.
+	 */
+	if (atomic_test_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
```

Comment:
When does TX_COMPLETE clear?

Reply:
See the pe_send_xxx_msg functions in usbc_pe_common.c

```
void pe_send_data_msg(const struct device *dev, const enum pd_packet_type type,
                        const enum pd_data_msg_type msg)
{
        struct usbc_port_data *data = dev->data;
        struct policy_engine *pe = data->pe;

        /* Clear any previous TX status before sending a new message */
        atomic_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE);
        prl_send_data_msg(dev, type, msg);
}
```